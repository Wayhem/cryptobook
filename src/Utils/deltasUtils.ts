import { XBTUSD_GROUPS_ENUM, ETHUSD_GROUPS_ENUM } from 'Models/constants/groups'
import Delta, { DeltaWithoutTotal } from 'Models/Delta'
import OrdersBook from 'Models/OrderBook'

function getClosestMultiple(x: number, y: number) {
  return Math.round(x / y) / (1 / y)
}

function isMultiple(x: number, y: number) {
  return getClosestMultiple(x, y) === x
}

const groupDeltasByNumber = (group: XBTUSD_GROUPS_ENUM | ETHUSD_GROUPS_ENUM, deltas: Delta[]) => {
  const groupedDeltas = [] as Delta[]
  if (Array.isArray(deltas) && deltas.length) {
    deltas.forEach(delta => {
      if (groupedDeltas.length === 0) {
        groupedDeltas.push([getClosestMultiple(delta[0], group), delta[1], delta[2]])
      } else if (isMultiple(delta[0], group) && delta[0] !== groupedDeltas[groupedDeltas.length - 1][0]) {
        groupedDeltas.push(delta)
      } else if (getClosestMultiple(delta[0], group) === groupedDeltas[groupedDeltas.length - 1][0]) {
        groupedDeltas[groupedDeltas.length - 1][1] += delta[1]
        groupedDeltas[groupedDeltas.length - 1][2] += delta[2]
      } else {
        groupedDeltas.push([getClosestMultiple(delta[0], group), delta[1], delta[2]])
      }
    })
  }
  return groupedDeltas
}

const getUpdatedDeltas = (newDeltas: DeltaWithoutTotal[], currentOrderBook: OrdersBook, isAsks: boolean): Delta[] => {
  const currentDeltas = isAsks ? [...currentOrderBook.asks] : [...currentOrderBook.bids]

  newDeltas.forEach((newDelta: DeltaWithoutTotal) => {
    if (newDelta[1] === 0) {
      for (const [currentDeltaIndex, currentDelta] of currentDeltas.entries()) {
        if (currentDelta[0] === newDelta[0]) {
          currentDeltas.splice(currentDeltaIndex, 1)
          break
        }
      }
    } else if (
      isAsks
        ? newDelta[0] > currentDeltas[currentDeltas.length - 1][0]
        : newDelta[0] < currentDeltas[currentDeltas.length - 1][0]
    ) {
      currentDeltas.push([newDelta[0], newDelta[1], newDelta[1]])
    } else if (isAsks ? newDelta[0] < currentDeltas[0][0] : newDelta[0] > currentDeltas[0][0]) {
      currentDeltas.unshift([newDelta[0], newDelta[1], newDelta[1]])
    } else {
      for (const [currentDeltaIndex, currentDelta] of currentDeltas.entries()) {
        if (currentDelta[0] === newDelta[0]) {
          currentDeltas[currentDeltaIndex][1] = newDelta[1]
          currentDeltas[currentDeltaIndex][2] += newDelta[1]
        } else if (
          isAsks
            ? currentDelta[0] > newDelta[0] && currentDeltas[currentDeltaIndex - 1][0] < newDelta[0]
            : currentDelta[0] < newDelta[0] && currentDeltas[currentDeltaIndex - 1][0] > newDelta[0]
        ) {
          currentDeltas.splice(currentDeltaIndex, 0, [newDelta[0], newDelta[1], newDelta[1]])
        }
      }
    }
  })

  return currentDeltas
}

export { groupDeltasByNumber, getUpdatedDeltas }
