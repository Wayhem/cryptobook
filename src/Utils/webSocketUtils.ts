import WSMessage from 'Models/WSMessage'
import ProductIds from 'Models/constants/productIds'
import WSMessageEvents from 'Models/constants/webSocketMessageEvents'
import FeedTypes from 'Models/constants/feedTypes'
import { XBTUSD_GROUPS_ENUM } from 'Models/constants/groups'
import Delta from 'Models/Delta'

const buildToggleWSMessage = (event: WSMessageEvents, productId: ProductIds): WSMessage => ({
  event,
  feed: FeedTypes.standard,
  product_ids: [productId],
})

function getClosestMultiple(x: number, y: number) {
  return Math.round(x / y) / (1 / y)
}

function isMultiple(x: number, y: number) {
  return getClosestMultiple(x, y) === x
}

const groupDeltasByNumber = (group: XBTUSD_GROUPS_ENUM, deltas: Delta[]) => {
  const groupedDeltas = [] as Delta[]
  if (Array.isArray(deltas) && deltas.length) {
    deltas.forEach(delta => {
      if (groupedDeltas.length === 0) {
        groupedDeltas.push([getClosestMultiple(delta[0], group), delta[1], delta[2]])
      } else if (isMultiple(delta[0], group) && delta[0] !== groupedDeltas[groupedDeltas.length - 1][0]) {
        groupedDeltas.push(delta)
      } else if (getClosestMultiple(delta[0], group) === groupedDeltas[groupedDeltas.length - 1][0]) {
        groupedDeltas[groupedDeltas.length - 1][1] += delta[1]
        groupedDeltas[groupedDeltas.length - 1][2] += delta[2]
      } else {
        groupedDeltas.push([getClosestMultiple(delta[0], group), delta[1], delta[2]])
      }
    })
  }
  return groupedDeltas
}

export { buildToggleWSMessage, groupDeltasByNumber }
