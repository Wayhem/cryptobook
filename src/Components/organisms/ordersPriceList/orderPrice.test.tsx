import OrderPrice from 'Components/organisms/ordersPriceList'
import Alignments from 'Models/constants/priceAlignments'
import { render } from 'Utils/test-utils'

describe('OrderPrice page component', () => {
  test('OrderPrice renders and matches snapshot', () => {
    const defaultProps = {
      alignment: Alignments.left,
      elements: [],
      biggestNumber: 0,
      color: 'red',
      colorBg: 'lightRed',
    }
    const { container } = render(<OrderPrice {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })
})
