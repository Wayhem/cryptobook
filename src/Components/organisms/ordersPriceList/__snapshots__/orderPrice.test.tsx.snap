// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderPrice page component OrderPrice renders and matches snapshot 1`] = `
<div>
  <div
    class="sc-eCstlR kjpSlN"
  >
    <div
      class="sc-bdfBQB dPOicM"
    >
      <div
        class="sc-dlfnuX iJDqXm"
      >
        price
      </div>
      <div
        class="sc-dlfnuX iJDqXm"
      >
        size
      </div>
      <div
        class="sc-dlfnuX iJDqXm"
      >
        total
      </div>
    </div>
  </div>
</div>
`;
